visallo/visallo
# Allow users to add items to ontology

Users with `ONTOLOGY_ADD` can now extend the ontology (in the current case, and all cases if they have `ONTOLOGY_PUBLISH`) to add new entity / edge types and properties.


![screen shot 2017-07-19 at 11 57 16 am](https://user-images.githubusercontent.com/7098/28379570-b98e684e-6c79-11e7-8316-6dcf56a2edad.png)
![screen shot 2017-07-19 at 11 57 39 am](https://user-images.githubusercontent.com/7098/28379569-b98c9c08-6c79-11e7-8e36-7e050b5d9a26.png)
![screen shot 2017-07-19 at 11 58 32 am](https://user-images.githubusercontent.com/7098/28379614-ebacca82-6c79-11e7-91e7-0166414a1bf1.png)
![screen shot 2017-07-19 at 11 59 04 am](https://user-images.githubusercontent.com/7098/28379567-b978516c-6c79-11e7-9e2a-c58653c90c68.png)

Testing Instructions: 
* Create concepts, relationships, properties. Must have `ONTOLOGY_ADD`
* Users without `ONTOLOGY_PUBLISH` should not see publish button on things that have schema changes (with a message)
* Publish. Must have `ONTOLOGY_PUBLISH`
* Publish with different user than the creator. Should see a warning about schema changes
* Test components that use new ontology selectors:
    * Saved Search
    * Structured Import
    * Property forms
    * Dashboard aggregation
    * Dashboard search card
    * polar dates
    * Statement form (drag entities in text)
    * Resolve Term
    * Add Related
    * File import (multiple)
    * requeue
    * admin ontology forms

Points of Regression: Deprecations when using ontologyRepository

CHANGELOG
Added: Users can dynamically add new concepts, edge types, and properties to the ontology using the ontology selection typeahead components. Users need to have `ONTOLOGY_ADD` to add unpublished items, and `ONTOLOGY_PUBLISH` to publish.
Changed: The ontology color of concepts is now shown in the graph product. Icons that are PNG (4 channel) with alpha will be used as a mask with the color.
Changed: Relationships are shown in hierarchy in dropdown selection lists.
Changed: BREAKING - All methods in `org.visallo.core.security.ACLProvider` now take a workspaceId to be used when working with sandboxed ontologies. All custom implementations must be updated to support the new method signatures.
Changed: Added validation to ensure that all DatatypeProperty's within an ontology have both `range` and at least one `domain` specified.
Deprecated: Several methods in OntologyRepository are deprecated because they now accept a workspaceId. Please check usages and pass workspace if appropriate to work with unpublished ontology items.
Deprecated: React was upgraded to 15.6.1 which deprecates `React.createClass` and `React.PropTypes`. For plugins that use those they need to `require(['create-react-class', 'prop-types'], (createReactClass, PropTypes) => {})` instead or you will receive console warnings. Visallo makes those modules available through requirejs.
